feat(Core): switch settings to TypeORM

kyranet opened the pull request:
Fixes #1048

favna commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -188,18 +186,33 @@ export default class extends SkyraCommand {
+				const toUpdate: TwitchStreamSubscriptionEntity[] = [];
+				const toDelete: TwitchStreamSubscriptionEntity[] = [];
+				for (const entry of entries) {
+					if (entry.guildIds.length === 1) {
```
It was `=== 0` before?

kyranet replied:
Because the entry was removed if it became empty, we optimized this flow by going from:

`remove from array` -> `check if array is empty` -> `delete from database`
(which comes with 1 write, 1 read, and 1 write)

to:

`get entries containing guild` -> `check if the array will be empty` -> `update/delete from database`
(which comes with 1 read and 1 write)

favna commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -188,18 +186,33 @@ export default class extends SkyraCommand {
-			const entries = subscriptions.reduce((accumulator, subscription) => accumulator + subscription[1].length, 0);
-			if (entries === 0) throw message.language.tget('COMMAND_TWITCHSUBSCRIPTION_RESET_EMPTY');
+			const subscription = subscriptions.reduce((accumulator, subscription) => accumulator + subscription[1].length, 0);
+			if (subscription === 0) throw message.language.tget('COMMAND_TWITCHSUBSCRIPTION_RESET_EMPTY');
```
```suggestion
			const subscriptions = subscriptions.reduce((accumulator, subscription) => accumulator + subscription[1].length, 0);
			if (subscriptions === 0) throw message.language.tget('COMMAND_TWITCHSUBSCRIPTION_RESET_EMPTY');
```

Don't remove the plural for arrays

favna commented on src/lib/util/Leaderboard.ts:
```diff
@@ -1,10 +1,12 @@
 /**
  * The Leaderboard singleton class in charge of storing local and global leaderboards
- * @version 2.0.0
+ * @version 2.1.0
```
Not that I mind but... huh?

kyranet replied:
Kind of versioning Skyra's sub-modules, lmao

favna commented on src/tasks/twitchRefreshSubscriptions.ts:
```diff
@@ -31,16 +32,13 @@ export default class extends Task {
 				// Queue the updating by pushing the promise into the promises array
-				promises.push(
-					this.client.twitch.subscriptionsStreamHandle(subscription.id, TwitchHooksAction.Subscribe).catch(error => this.client.emit(Events.Wtf, error)),
-					this.client.queries.upsertTwitchStreamSubscription(subscription.id).catch(error => this.client.emit(Events.Wtf, error))
```
This needs to be restored. `subscriptionStreamHandle` only resubs on Twitch but it doesn't update the entry in the database.

kyranet replied:
But the entry is already there, as you see in a few lines above where we fetch it.

favna replied:
Yes and the expireAt has to be updated to the new date otherwise it'll think it has to refresh it for another 10 days the next time the task runs

kyranet replied:
They should be updated using a TypeORM transaction, iirc using `createQueryRunner()` (so we can use `Promise.all`).
- The previous code was doing an UPSERT operation for each entry even though they already existed, here we only need to update the field and do a batch update.
- The other way to do it is by using `update()` which takes a condition (can be used to select all streamer IDs) and the new date for the `refreshAt`, this way we do a single query as opposed to making multiple queries within a transaction.

favna reviewed (changes requested):
<snip>

favna committed:
feat(*): migrate settings to TypeORM

- diasble grafana data
- add typeorm resync script
- setup  database error logging
- fix typeorm script
- chore(ORM): add migration file (unfinished)
- fix(ORM): of course I did not forget to rename the file
- fix(ORM): resolved some TODOs, fixed many relationships
- fix(ORM): use .js instead of .json, use naming strategy
- cleanup(ORM): refactored entities
- refactor(*): use TypeORM everywhere but guilds
- fix(*): update everything to DbSet.fetchColor
- refactor(ORM): rename keys
- docs: update @vladfrangu as a contributor
- refactor(*): apply suggestions from code review
- test: add test for getColor
- test: cover getImage and getAttachment
- chore: organize all imports
- regen lockfile
- fix resync script
- fix guild entity
- fixes
- fix interop with klasa client & guild schema
- fix twitch upsert subscription
- test: use fs.promises.readfile
- sql highlight inline sql
- build: migrate to @skyra/eslint-config
- temp(User): attempt to fix UserEntity#save's raceconditions
- fix(ORM): doesn't work, and will never work, fuck TypeORM
- update lockfile
- feat(ORM): add locking for UserRepository
- fix(marriages): resolved all bugs regarding marriages
- fix(giveaway): resolved all bugs regarding giveaways
- fix(social): resolved leaderboard bug
- write migrations for ClientEntity and DashboardUserEntity
- feat(core): added Schedule
- src: make check more readable (and accurate) (#1067)
- fix(moderation,schedule): resolved couple of bugs
- schedule: set pgsql defaults
- write migration for schedules
- fix schedule createTable
- organize imports
- I just evaluated random code in my console
- Too lazy to write descriptive message
- fix(*): resolved all requested changes
- banneentity: add checks
- add migrations banners -> banner
- add checks for guild,giveaway, and command_counter
- migrations for giveaway
- organize some imports
- refix typeorm schema
- finish guild migrations
- finish marriage tweaks
- fix guild migrations
- fix migrations revert
- member migrations
- moderation entity fixes
- migrate moderations & greatly optimize for speed
- fix moderation revert

Co-authored-by: Vlad Frangu <kingdgrizzle@gmail.com>
Co-authored-by: Jeroen Claassens <support@favware.tech>
Co-authored-by: Antonio Román <kyradiscord@gmail.com>

vladfrangu commented on src/commands/Moderation/Utilities/reason.ts:
```diff
@@ -28,12 +29,18 @@ export default class extends SkyraCommand {
+			.where('guild_id = :guild', { guild: message.guild!.id })
+			.andWhere('case_id IN (:...ids)', { ids: [...entries.keys()] })
+			.set({ reason, imageURL })
+			.returning('*')
```
Why do you need to return anything?

kyranet replied:
```suggestion
```

vladfrangu commented on src/commands/Moderation/Utilities/reason.ts:
```diff
@@ -28,12 +29,18 @@ export default class extends SkyraCommand {
 		const imageURL = getImage(message);
-		await this.client.queries.updateModerationLogReasonBulk(message.guild!.id, entries.map(ml => ml.case!), reason);
+		const connection = await DbSet.connect();
+		await connection.moderations.createQueryBuilder()
```
```suggestion
		const { moderations } = await DbSet.connect();
		await moderations.createQueryBuilder()
```

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -24,29 +27,66 @@ export default class extends SkyraCommand {
+			settings.points = amount;
+			await settings.save();
+		} else {
+			oldValue = 0;
```
You can default the oldValue to 0

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -24,29 +27,66 @@ export default class extends SkyraCommand {
 	public async add(message: KlasaMessage, [user, amount]: [KlasaUser, number]) {
-		const newAmount = await this.client.queries.upsertIncrementMemberSettings(message.guild!.id, user.id, amount);
-		return message.sendLocale('COMMAND_SOCIAL_ADD', [user.username, newAmount, amount]);
+		const connection = await DbSet.connect();
```
Destructure `members` from the object
```suggestion
		const { members } = await DbSet.connect();
```

vladfrangu commented on src/commands/Social/mylevel.ts:
```diff
@@ -18,8 +20,8 @@ export default class extends SkyraCommand {
 	public async run(message: KlasaMessage, [user = message.author]: [KlasaUser]) {
-		const memberSettings = await this.client.queries.fetchMemberSettings(message.guild!.id, user.id);
-		const memberPoints = memberSettings ? memberSettings.point_count : 0;
+		const memberSettings = await (await DbSet.connect()).members.findOne({ where: { userID: user.id, guildID: message.guild!.id }, cache: Time.Minute * 15 });
```
Consistency with the rest of the files, please
```suggestion
		const { members } = await DbSet.connect();
		const memberSettings = await members.findOne({ where: { userID: user.id, guildID: message.guild!.id }, cache: Time.Minute * 15 });
```

vladfrangu commented on src/commands/Starboard/star.ts:
```diff
@@ -1,39 +1,52 @@
-			: this.client.queries.fetchStarRandom(message.guild!.id, min));
+		const minimum = message.guild!.settings.get(GuildSettings.Starboard.Minimum);
+		const connection = await DbSet.connect();
+		const qb = connection.starboards.createQueryBuilder()
```
```suggestion
		const { starboards } = await DbSet.connect();
		const qb = starboards.createQueryBuilder()
```

vladfrangu commented on src/commands/Starboard/star.ts:
```diff
@@ -52,44 +65,52 @@ export default class extends SkyraCommand {
-			: this.client.queries.fetchStars(message.guild!.id, min));
+		const minimum = message.guild!.settings.get(GuildSettings.Starboard.Minimum);
+		const connection = await DbSet.connect();
+		const qb = connection.starboards.createQueryBuilder()
```
```suggestion
		const { starboards } = await DbSet.connect();
		const qb = starboards.createQueryBuilder()
```

vladfrangu commented on src/commands/Suggestions/resolvesuggestion.ts:
```diff
@@ -92,19 +93,27 @@ export default class extends SkyraCommand {
+
+			// Retrieve the suggestion data
+			const connection = await DbSet.connect();
+			const suggestionData = await connection.suggestions.findOne({ id, guildID: message.guild!.id });
```
```suggestion
			const { suggestions } = await DbSet.connect();
			const suggestionData = await suggestions.findOne({ id, guildID: message.guild!.id });
```

vladfrangu commented on src/commands/Suggestions/suggest.ts:
```diff
@@ -51,11 +52,12 @@ export default class extends SkyraCommand {
 		// Commit the suggestion to the DB
-		await this.client.queries.insertSuggestion({
+		const connection = await DbSet.connect();
+		await connection.suggestions.insert({
```
```suggestion
		const { suggestions } = await DbSet.connect();
		await suggestions.insert({
```

vladfrangu commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -184,40 +188,50 @@ export default class extends SkyraCommand {
-			if (updated.length !== 0) await this.client.queries.deleteTwitchStreamSubscriptions(pendingToDelete);
-			return message.sendLocale('COMMAND_TWITCHSUBSCRIPTION_RESET_SUCCESS', [entries]);
+			const connection = await DbSet.connect();
+			await connection.twitchStreamSubscriptions.manager.transaction(async em => {
```
```suggestion
			const { twitchStreamSubscriptions } = await DbSet.connect();
			await twitchStreamSubscriptions.manager.transaction(async em => {
```

vladfrangu commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -285,4 +299,38 @@ export default class extends SkyraCommand {
 
+	private async upsertSubscription(guild: Guild, streamer: Streamer) {
+		const connection = await DbSet.connect();
+		const results = await connection.twitchStreamSubscriptions.createQueryBuilder()
```
```suggestion
		const { twitchStreamSubscriptions } = await DbSet.connect();
		const results = await twitchStreamSubscriptions.createQueryBuilder()
```

vladfrangu commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -285,4 +299,38 @@ export default class extends SkyraCommand {
+
+	private async removeSubscription(guild: Guild, streamer: Streamer) {
+		const connection = await DbSet.connect();
+		const subscription = await connection.twitchStreamSubscriptions.findOne({ id: streamer.id });
```
```suggestion
		const { twitchStreamSubscriptions } = await DbSet.connect();
		const subscription = await twitchStreamSubscriptions.findOne({ id: streamer.id });
```

vladfrangu commented on src/events/klasaReady.ts:
```diff
@@ -65,4 +69,15 @@ export default class extends Event {
+			.filter(command => !entries.some(entry => entry.id === command.name))
+			.map(command => new CommandCounterEntity().setID(command.name));
+
+		if (missing.length) await connection.commandCounters.insert(missing);
```
```suggestion
		const { commandCounters } = await DbSet.connect();
		const entries = await commandCounters.find();

		const missing = this.client.commands
			.filter(command => !entries.some(entry => entry.id === command.name))
			.map(command => new CommandCounterEntity().setID(command.name));

		if (missing.length) await commandCounters.insert(missing);
```

vladfrangu commented on src/events/rawChannelDelete.ts:
```diff
@@ -0,0 +1,56 @@
+		// Delete entries from starboard if it exists
+		try {
+			const connection = await DbSet.connect();
+			const results = await connection.starboards.createQueryBuilder()
```
```suggestion
			const { starboards } = await DbSet.connect();
			const results = await starboards.createQueryBuilder()
```

vladfrangu commented on src/events/rawChannelDelete.ts:
```diff
@@ -0,0 +1,56 @@
+				return;
+			}
+			await api(this.client).channels(channel).messages['bulk-delete']
+				.post({ data: { messages: filteredResults }, reason: 'Starboard Management: Message Deleted' })
```
This will error for any messages older than 14 days. Might wanna change the logic

vladfrangu commented on src/events/rawMessageDelete.ts:
```diff
@@ -24,7 +25,16 @@ export default class extends Event {
 		try {
-			const result = await this.client.queries.deleteStarReturning(data.guild_id, data.id);
+			const connection = await DbSet.connect();
+			const results = await connection.starboards.createQueryBuilder()
```
```suggestion
			const { starboards } = await DbSet.connect();
			const results = await starboards.createQueryBuilder()
```

vladfrangu commented on src/events/rawMessageDeleteBulk.ts:
```diff
@@ -16,17 +17,22 @@ export default class extends Event {
 		try {
-			const results = await this.client.queries.deleteStarsReturning(data.guild_id, data.ids);
+			const connection = await DbSet.connect();
+			const results = await connection.starboards.createQueryBuilder()
```
```suggestion
			const { starboards } = await DbSet.connect();
			const results = await starboards.createQueryBuilder()
```

vladfrangu commented on src/events/rawMessageReactionRemoveAll.ts:
```diff
@@ -18,14 +19,23 @@ export default class extends Event {
 		try {
-			const results = await this.client.queries.deleteStarReturning(data.guild_id, data.message_id);
+			const connection = await DbSet.connect();
+			const results = await connection.starboards.createQueryBuilder()
```
```suggestion
			const { starboards } = await DbSet.connect();
			const results = await starboards.createQueryBuilder()
```

vladfrangu commented on src/events/rawMessageReactionRemoveEmoji.ts:
```diff
@@ -20,14 +21,23 @@ export default class extends Event {
 		try {
-			const results = await this.client.queries.deleteStarReturning(data.guild_id, data.message_id);
+			const connection = await DbSet.connect();
+			const results = await connection.starboards.createQueryBuilder()
```
```suggestion
			const { starboards } = await DbSet.connect();
			const results = await starboards.createQueryBuilder()
```

vladfrangu commented on src/events/twitch/twitchStreamOffline.ts:
```diff
@@ -10,11 +11,11 @@ export default class extends Event {
 		// Fetch the streamer, and if it could not be found, return error.
-		const streamer = await this.client.queries.fetchTwitchStreamSubscription(data.id);
-		if (streamer === null) return response.error('No streamer could be found in the database.');
+		const streamer = await (await DbSet.connect()).twitchStreamSubscriptions.findOne({ id: data.id });
```
```suggestion
		const { twitchStreamSubscriptions } = await DbSet.connect();
		const streamer = await twitchStreamSubscriptions.findOne({ id: data.id });
```

vladfrangu commented on src/events/twitch/twitchStreamOnline.ts:
```diff
@@ -17,12 +18,12 @@ export default class extends Event {
 		// Fetch the streamer, and if it could not be found, return error.
-		const streamer = await this.client.queries.fetchTwitchStreamSubscription(data.user_id);
-		if (streamer === null) return response.error('No streamer could be found in the database.');
+		const streamer = await (await DbSet.connect()).twitchStreamSubscriptions.findOne({ id: data.user_id });
```
```suggestion
		const { twitchStreamSubscriptions } = await DbSet.connect();
		const streamer = await twitchStreamSubscriptions.findOne({ id: data.user_id });
```

vladfrangu commented on src/finalizers/commandUseCounter.ts:
```diff
@@ -1,9 +1,11 @@
 	public async run(_message: KlasaMessage, command: Command) {
-		await this.client.queries.insertCommandUseCounter(command.name);
+		const connection = await DbSet.connect();
+		await connection.commandCounters.increment({ id: command.name }, 'uses', 1);
```
```suggestion
		const { commandCounters } = await DbSet.connect();
		await commandCounters.increment({ id: command.name }, 'uses', 1);
```

vladfrangu commented on src/lib/orm/entities/CommandCounterEntity.ts:
```diff
@@ -0,0 +1,17 @@
+	@PrimaryColumn('varchar', { length: 32 })
+	public id: string = null!;
+
+	@Column('integer', { 'default': () => '0' })
```
Why is the default not a number?

vladfrangu commented on src/lib/orm/entities/CommandCounterEntity.ts:
```diff
@@ -0,0 +1,17 @@
+export class CommandCounterEntity extends BaseEntity {
+
+	@PrimaryColumn('varchar', { length: 32 })
+	public id: string = null!;
```
Why is this defaulted to a "non-null null"?

vladfrangu commented on src/lib/orm/entities/GuildEntity.ts:
```diff
@@ -0,0 +1,505 @@
+export class GuildEntity extends BaseEntity {
+
+	@PrimaryColumn('varchar', { name: 'id', length: 19 })
+	public id?: string;
```
id cannot be undefined

vladfrangu commented on src/lib/orm/entities/GuildEntity.ts:
```diff
@@ -0,0 +1,505 @@
+import { BaseEntity, Column, Entity, PrimaryColumn } from 'typeorm';
+
+@Entity('guilds', { schema: 'public' })
+export class GuildEntity extends BaseEntity {
```
Since I honestly can't be asked to write a review for everything, nested props can use https://typeorm.io/#/embedded-entities for a cleaner usage (for example `permissions.users` instead of `permissionsUsers`)

kyranet replied:
Yeah but guild settings aren't ported over for now. This is only there to provide migrations and stuff.

vladfrangu commented on src/lib/orm/entities/MemberEntity.ts:
```diff
@@ -0,0 +1,16 @@
+export class MemberEntity extends BaseEntity {
+
+	@PrimaryColumn('varchar', { length: 19 })
+	public guildID?: string;
```
Neither guildID nor userID can be undefined

vladfrangu commented on src/lib/orm/entities/MemberEntity.ts:
```diff
@@ -0,0 +1,16 @@
+	public userID?: string;
+
+	@Column('bigint', { 'default': () => '0', 'transformer': kBigIntTransformer })
+	public points = 0;
```
Since this is a bigint supposedly..shouldn't this (redundant) `= 0` be `= 0n`?

kyranet replied:
It uses bigint, yes, but because `2,147,483,648` (integer's max) isn't enough, `9,007,199,254,740,991` is. Either way if this breaks I can switch it to a bigint, it's just not a need for now, since those amounts are vastly high.

Switching it to a `bigint` type will make the maximum `9,223,372,036,854,775,808`, which is approximately 1000 times `number`'s limit.

vladfrangu commented on src/lib/orm/entities/RpgBattleEntity.ts:
```diff
@@ -0,0 +1,54 @@
+export class RpgBattleEntity extends BaseEntity {
+
+	@PrimaryGeneratedColumn({ type: 'bigint' })
+	public id?: string;
```
id cannot be undefined

vladfrangu commented on src/lib/orm/entities/RpgClassEntity.ts:
```diff
@@ -0,0 +1,31 @@
+export class RpgClassEntity extends BaseEntity {
+
+	@PrimaryGeneratedColumn({ type: 'integer' })
+	public id?: number;
```
id cannot be undefined

vladfrangu commented on src/lib/orm/entities/RpgGuildEntity.ts:
```diff
@@ -0,0 +1,46 @@
+export class RpgGuildEntity extends BaseEntity {
+
+	@PrimaryGeneratedColumn({ type: 'integer' })
+	public id?: number;
```
id cannot be undefined

vladfrangu commented on src/lib/orm/entities/RpgGuildRankEntity.ts:
```diff
@@ -0,0 +1,21 @@
+export class RpgGuildRankEntity extends BaseEntity {
+
+	@PrimaryGeneratedColumn({ type: 'integer', name: 'id' })
+	public id?: number;
```
id cannot be undefined

vladfrangu commented on src/lib/orm/entities/RpgItemEntity.ts:
```diff
@@ -0,0 +1,47 @@
+export class RpgItemEntity extends BaseEntity {
+
+	@PrimaryGeneratedColumn({ type: 'integer' })
+	public id?: number;
```
id cannot be undefined

vladfrangu commented on src/lib/orm/entities/RpgUserItemEntity.ts:
```diff
@@ -0,0 +1,32 @@
+export class RpgUserItemEntity extends BaseEntity {
+
+	@PrimaryGeneratedColumn({ type: 'bigint' })
+	public id?: string;
```
id cannot be undefined

type: 'bigint' but field type string?

kyranet replied:
`PrimaryGeneratedColumn` doesn't support `transformer`, and `pg` returns `bigint` as `string`s.

vladfrangu commented on src/lib/orm/entities/UserEntity.ts:
```diff
@@ -0,0 +1,42 @@
+	public moderationDM = true;
+
+	@Column('bigint', { 'default': () => '0', 'transformer': kBigIntTransformer })
+	public money = 0;
```
shouldn't `= 0` be `= 0n`?

vladfrangu commented on src/lib/orm/entities/UserProfileEntity.ts:
```diff
@@ -0,0 +1,36 @@
+	public color = 0x000000;
+
+	@Column('bigint', { 'default': () => '0', 'transformer': kBigIntTransformer })
+	public vault = 0;
```
Shouldn't `= 0` be `= 0n`?

vladfrangu commented on src/lib/structures/DbSet.ts:
```diff
@@ -0,0 +1,130 @@
+import type { Connection } from 'typeorm';
+import type { Message } from 'discord.js';
+
+
```
```suggestion

import type { Connection } from 'typeorm';
import type { Message } from 'discord.js';

```

favna replied:
We don't really space out type imports because they get remove by VSCode's organize imports anyway.

vladfrangu commented on src/lib/structures/GiveawayManager.ts:
```diff
@@ -1,27 +1,24 @@
-			? this.client.queries.fetchGiveawaysFromGuilds([...this.client.guilds.keys()])
-			: this.client.providers.default!.getAll(Databases.Giveaway)) as RawGiveawaySettings[];
+		const connection = await DbSet.connect();
+		const qb = connection.giveaways.createQueryBuilder().select();
```
```suggestion
		const { giveaways } = await DbSet.connect();
		const qb = giveaways.createQueryBuilder().select();
```

vladfrangu commented on src/lib/structures/GiveawayManager.ts:
```diff
@@ -1,27 +1,24 @@
+		if (this.client.shard) qb.where('guild_id IN (:...ids)', { ids: [...this.client.guilds.keys()] });
 
-		for (const entry of entries) this.add(entry);
+		for (const entry of await qb.getMany()) this.insert(entry.setup(this));
```
Is there a `for await of`?

vladfrangu commented on src/lib/structures/ModerationManager.ts:
```diff
@@ -69,51 +70,54 @@ export class ModerationManager extends Collection<number, ModerationManagerEntry
 
 		if (super.size !== this._count) {
-			this._cache(await this.guild.client.queries.fetchModerationLogsByGuild(this.guild.id), CacheActions.Fetch);
+			this._cache(await (await DbSet.connect()).moderations.find({ where: { guildID: this.guild.id } }), CacheActions.Fetch);
```
I simply can't with seeing the `await (await)` chains. Please connect once at the top of the function and re-use that

vladfrangu commented on src/lib/structures/ModerationManager.ts:
```diff
@@ -69,51 +70,54 @@ export class ModerationManager extends Collection<number, ModerationManagerEntry
 	public async count() {
 		if (this._count === null) {
-			this._count = await this.client.queries.fetchModerationCurrentCaseID(this.guild.id);
+			this._count = await (await DbSet.connect()).moderations.count({ where: { guildID: this.guild.id } });
```
It pains me to see this... Is there really no better way to avoid `await (await)`?

vladfrangu commented on src/lib/structures/StarboardManager.ts:
```diff
@@ -76,11 +70,19 @@ export class StarboardManager extends Collection<string, StarboardMessage> {
-			if (starboardMessage.users.size) starboardMessage.users.delete(userID);
-			return starboardMessage;
+			const connection = await DbSet.connect();
+			const previous = await connection.starboards.findOne({ where: { guildID: this.guild.id, messageID } });
```
```suggestion
			const { starboards } = await DbSet.connect();
			const previous = await starboards.findOne({ where: { guildID: this.guild.id, messageID } });
```

favna commented on src/lib/util/util.ts:
```diff
@@ -403,8 +403,15 @@ export function getImage(message: Message): string | null {
+	const user = await userProfiles.findOne(message.author.id);
+
+	return user?.color || message.member?.displayColor || BrandingColors.Primary;
+}
```
This should get a unit test. Mock out a message:
```ts
const fakeMessage = {
  author: {
    id: '132456978'
  }
}
```

And mock `DbSet.connect()`: `jest.spyOn(DbSet, 'connect').mockImplementation(() => fakeUserProfiles)` (create a fakeUserProfiles constant which has `findOne` method)

Then ensure all branches are covered.

favna replied:
Otherwise we can't merge because coveralls makes CI red (and rightfully so)

kyranet replied:
This code is outdated, fetchColor is not longer in util but in DbSet.

vladfrangu commented on src/lib/util/Leaderboard.ts:
```diff
@@ -111,11 +113,18 @@ export class Leaderboard {
 
 	private async createMemberSyncHandle(guild: string) {
-		const data = await this.client.queries.fetchLeaderboardLocal(guild);
+		const connection = await DbSet.connect();
```
Why not set a private `#connection` property? Otherwise..

vladfrangu commented on src/lib/util/Leaderboard.ts:
```diff
@@ -111,11 +113,18 @@ export class Leaderboard {
 	private async createMemberSyncHandle(guild: string) {
-		const data = await this.client.queries.fetchLeaderboardLocal(guild);
+		const connection = await DbSet.connect();
+		const data = await connection.users.createQueryBuilder()
```
```suggestion
		const { users } = await DbSet.connect();
		const data = await users.createQueryBuilder()
```

vladfrangu commented on src/lib/util/Leaderboard.ts:
```diff
@@ -148,14 +157,21 @@ export class Leaderboard {
+		const connection = await DbSet.connect();
 		// Get the sorted data from the db
-		const data = await this.client.queries.fetchLeaderboardGlobal();
+		const data = await connection.users.createQueryBuilder()
```
```suggestion
		const data = await users.createQueryBuilder()
```

vladfrangu commented on src/lib/util/Leaderboard.ts:
```diff
@@ -148,14 +157,21 @@ export class Leaderboard {
 	}
 
 	private async createUserSyncHandle() {
+		const connection = await DbSet.connect();
```
Ditto or
```suggestion
		const { users } = await DbSet.connect();
```

vladfrangu commented on src/lib/util/util.ts:
```diff
@@ -403,7 +403,7 @@ export function getImage(message: Message): string | null {
 }
 
 export function getColor(message: Message) {
```
Isn't this function redundant now?

vladfrangu commented on src/lib/util/util.ts:
```diff
@@ -594,6 +594,8 @@ export function bidirectionalReplace<T>(regex: RegExp, content: string, options:
 	return results;
 }
 
+export const kBigIntTransformer: ValueTransformer = { from: Number, to: String };
```
Why not to BigInt? You're on nodejs 14 after all

vladfrangu commented on src/monitors/Moderation/capitalsFilter.ts:
```diff
@@ -55,7 +55,7 @@ export default class extends ModerationMonitor {
 	protected async onDelete(message: KlasaMessage, value: number) {
 		floatPromise(this, message.nuke());
-		if (value > 25 && (await message.author.settings.sync()).get(UserSettings.ModerationDM)) {
+		if (value > 25 && await (await DbSet.connect()).users.moderationDirectMessageEnabled(message.author.id)) {
```
```suggestion
		const { users } = await DbSet.connect();
		if (value > 25 && await users.moderationDirectMessageEnabled(message.author.id)) {
```

vladfrangu commented on src/monitors/Moderation/wordFilter.ts:
```diff
@@ -36,7 +36,7 @@ export default class extends ModerationMonitor {
 	protected async onDelete(message: KlasaMessage, value: FilterResults) {
 		floatPromise(this, message.nuke());
-		if (message.content.length > 25 && (await message.author.settings.sync()).get(UserSettings.ModerationDM)) {
+		if (message.content.length > 25 && await (await DbSet.connect()).users.moderationDirectMessageEnabled(message.author.id)) {
```
```suggestion
		const { users } = await DbSet.connect();
		if (message.content.length > 25 && await users.moderationDirectMessageEnabled(message.author.id)) {
```

vladfrangu commented on src/routes/oauth/oauthCallback.ts:
```diff
@@ -56,12 +58,21 @@ export default class extends Route {
+			created.accessToken = body.access_token;
+			created.refreshToken = body.refresh_token;
+			await connection.dashboardUsers.insert(created);
+		}
```
```suggestion
		const { dashboardUsers } = await DbSet.connect();
		const existing = await dashboardUsers.findOne({ where: { id: user.id }, cache: Time.Minute });
		if (existing) {
			existing.expiresAt = new Date(Date.now() + (body.expires_in * Time.Second));
			existing.accessToken = body.access_token;
			existing.refreshToken = body.refresh_token;
			await existing.save();
		} else {
			const created = new DashboardUserEntity();
			created.id = user.id;
			created.expiresAt = new Date(Date.now() + (body.expires_in * Time.Second));
			created.accessToken = body.access_token;
			created.refreshToken = body.refresh_token;
			await dashboardUsers.insert(created);
		}
```

vladfrangu commented on src/routes/oauth/oauthUser.ts:
```diff
@@ -35,13 +34,23 @@ export default class extends Route {
-			let data = await this.client.queries.fetchDashboardUser(request.auth!.user_id);
-			if (data === null) return response.error(401);
+			const connection = await DbSet.connect();
+			const data = await connection.dashboardUsers.findOne({ where: { id: request.auth!.user_id }, cache: Time.Minute });
```
```suggestion
			const { dashboardUsers } = await DbSet.connect();
			const data = await dashboardUsers.findOne({ where: { id: request.auth!.user_id }, cache: Time.Minute });
```

vladfrangu commented on src/tasks/twitchRefreshSubscriptions.ts:
```diff
@@ -14,7 +15,7 @@ export default class extends Task {
 
 		// Retrieve all the Twitch subscriptions
-		const allSubscriptions = await this.client.queries.fetchAllTwitchStreams();
+		const allSubscriptions = await (await DbSet.connect()).twitchStreamSubscriptions.find();
```
Please..no more `await (await)`. Extract the variable, and pass it down to `updateEntries` as well, skip one connect call. OR HELL, pass the `twitchStreamSubscriptions` repository over

vladfrangu reviewed (changes requested):
<snip>

favna commented on src/lib/orm/migrations/Type600.ts:
```diff
@@ -0,0 +1,91 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
```
If the idea is that we run this migration on prod then it needs `IF NOT EXISTS` for sure for the table queries but other than that it's also far from complete because it's missing all the data migrations (INSERT INTO statements)

favna commented on src/commands/Social/daily.ts:
```diff
@@ -1,37 +1,37 @@
-		const time = message.author.settings.get(UserSettings.TimeDaily);
+
+		const set = await DbSet.connect();
+		const settings = await set.users.ensureCooldowns(message.author.id);
```
```suggestion
		const connection = await DbSet.connect();
		const settings = await connection.users.ensureCooldowns(message.author.id);
```

favna commented on src/commands/Social/daily.ts:
```diff
@@ -1,37 +1,37 @@
+		if (!settings.cooldowns.daily || settings.cooldowns.daily.getTime() <= now) {
 			return message.sendLocale('COMMAND_DAILY_TIME_SUCCESS',
-				[await this.claimDaily(message, now + DAILY_PERIOD)]);
+				[await this.claimDaily(message, set, settings, now + DAILY_PERIOD)]);
```
```suggestion
				[await this.claimDaily(message, connection, settings, now + DAILY_PERIOD)]);
```

favna commented on src/commands/Social/daily.ts:
```diff
@@ -42,27 +42,27 @@ export default class extends SkyraCommand {
 		// The user accepted the grace period
 		return message.sendLocale('COMMAND_DAILY_GRACE_ACCEPTED', [
-			await this.claimDaily(message, now + remaining + DAILY_PERIOD),
+			await this.claimDaily(message, set, settings, now + remaining + DAILY_PERIOD),
```
```suggestion
			await this.claimDaily(message, connection, settings, now + remaining + DAILY_PERIOD),
```

favna commented on src/commands/Social/daily.ts:
```diff
@@ -42,27 +42,27 @@ export default class extends SkyraCommand {
-		const money = this.calculateDailies(message);
-		const currentMoney = await message.author.settings.get(UserSettings.Money);
+	private async claimDaily(message: KlasaMessage, set: DbSet, settings: UserEntity, nextTime: number) {
+		const money = this.calculateDailies(message, await set.clients.ensure(), settings);
```
```suggestion
	private async claimDaily(message: KlasaMessage, connection: DbSet, settings: UserEntity, nextTime: number) {
		const money = this.calculateDailies(message, await connection.clients.ensure(), settings);
```

vladfrangu commented on tests/lib/util/util.test.ts:
```diff
@@ -288,6 +290,130 @@ describe('Utils', () => {
+		test('GIVEN message w/ attachments w/ image w/o proxyURL attachment THEN returns url', async () => {
+			const filePath = resolve(__dirname, '..', '..', '..', 'tests', 'mocks', 'image.png');
+			const readStream = createReadStream(filePath);
+			const buffer = await utils.streamToBuffer(readStream);
```
Any reason you don't directly fs.promises.readFile?

kyranet replied:
It's... the point of testing ``streamToBuffer``...

vladfrangu replied:
No...you're testing getImage, which has nothing to do with streamToBuffer?

vladfrangu commented on tests/lib/util/util.test.ts:
```diff
@@ -288,6 +290,130 @@ describe('Utils', () => {
+		test('GIVEN message w/ attachments w/ image w/ proxyURL attachment THEN returns url', async () => {
+			const filePath = resolve(__dirname, '..', '..', '..', 'tests', 'mocks', 'image.png');
+			const readStream = createReadStream(filePath);
+			const buffer = await utils.streamToBuffer(readStream);
```
Ditto

vladfrangu commented on tests/lib/util/util.test.ts:
```diff
@@ -288,6 +290,130 @@ describe('Utils', () => {
+		test('GIVEN message w/ attachments w/o image attachment THEN passes through to embed checking', async () => {
+			const filePath = resolve(__dirname, '..', '..', '..', 'tests', 'mocks', 'image.png');
+			const readStream = createReadStream(filePath);
+			const buffer = await utils.streamToBuffer(readStream);
```
Ditto!

kyranet committed:
feat(orm): add user migration

kyranet committed:
fix(orm): remove sql schema

kyranet committed:
feat(orm): add missing Check in StarboardEntity

kyranet committed:
refactor(orm): use renameTable instead of SQL for banners migration

kyranet committed:
feat(orm): add suggestion migration

kyranet committed:
feat(orm): add twitch stream subscription migration

kyranet committed:
feat(orm): added checks for all RPG tables

kyranet committed:
refactor(orm): use json_populate_recordset for giveaway migration

kyranet committed:
refactor(orm): avoid second query for user_profile migration

kyranet committed:
feat(orm): add migration for RPG tables

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -24,29 +27,65 @@ export default class extends SkyraCommand {
+		const { members } = await DbSet.connect();
+		const settings = await members.findOne({ where: { userID: user.id, guildID: message.guild!.id }, cache: Time.Minute * 15 });
+		if (settings) {
+			const newAmount = Number(settings.points) + amount;
```
Why the Number conversion here?

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -24,29 +27,65 @@ export default class extends SkyraCommand {
 	public async remove(message: KlasaMessage, [user, amount]: [KlasaUser, number]) {
-		const newAmount = await this.client.queries.upsertDecrementMemberSettings(message.guild!.id, user.id, amount);
+		const connection = await DbSet.connect();
+		const settings = await connection.members.findOne({ where: { userID: user.id, guildID: message.guild!.id }, cache: Time.Minute * 15 });
```
```suggestion
		const { members } = await DbSet.connect();
		const settings = await members.findOne({ where: { userID: user.id, guildID: message.guild!.id }, cache: Time.Minute * 15 });
```

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -24,29 +27,65 @@ export default class extends SkyraCommand {
+		const settings = await connection.members.findOne({ where: { userID: user.id, guildID: message.guild!.id }, cache: Time.Minute * 15 });
+		if (!settings) throw message.language.tget('COMMAND_SOCIAL_MEMBER_NOTEXISTS');
+
+		const newAmount = Math.max(Number(settings.points) - amount, 0);
```
Why the cast to Number

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -24,29 +27,65 @@ export default class extends SkyraCommand {
+			created.guildID = message.guild!.id;
+			created.points = amount;
+			await connection.members.insert(created);
+		}
```
```suggestion
		const { members } = await DbSet.connect();
		const settings = await members.findOne({ where: { userID: user.id, guildID: message.guild!.id }, cache: Time.Minute * 15 });
		let oldValue = 0;
		if (settings) {
			oldValue = Number(settings.points);
			settings.points = amount;
			await settings.save();
		} else {
			const created = new MemberEntity();
			created.userID = user.id;
			created.guildID = message.guild!.id;
			created.points = amount;
			await members.insert(created);
		}
```

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -24,29 +27,65 @@ export default class extends SkyraCommand {
+		const settings = await connection.members.findOne({ where: { userID: user.id, guildID: message.guild!.id }, cache: Time.Minute * 15 });
+		let oldValue = 0;
+		if (settings) {
+			oldValue = Number(settings.points);
```
Why cast to Number

vladfrangu commented on src/commands/Social/Member Management/social.ts:
```diff
@@ -24,29 +27,65 @@ export default class extends SkyraCommand {
 
 	public async reset(message: KlasaMessage, [user]: [KlasaUser]) {
-		await this.client.queries.deleteMemberSettings(message.guild!.id, user.id);
+		await (await DbSet.connect()).members.delete({ userID: user.id, guildID: message.guild!.id });
```
Again, please try to stick away from `await (await...`

vladfrangu commented on src/commands/Suggestions/resolvesuggestion.ts:
```diff
@@ -92,19 +93,27 @@ export default class extends SkyraCommand {
+
+			const channel = this.client.channels.get(channelID) as TextChannel;
+			const suggestionMessage = await channel.messages.fetch(suggestionData.messageID);
 			if (typeof suggestionMessage === 'undefined') {
```
If check will never return true, since fetch will throw an error

kyranet replied:
Git blames @cfanoulis: b1ad290402009ebd6e1faacafdf8ab64172c9c7b

kyranet committed:
fix(*): resolved requested changes

Quantumlyy commented on src/commands/Admin/roleset.ts:
```diff
@@ -16,15 +16,29 @@ import { ApplyOptions } from '@skyra/decorators';
 	usage: '<add|remove|reset|list|auto:default> (name:name) (role:rolenames)',
 	usageDelim: ' '
 })
+@CreateResolvers([
+	[
+		'name', (arg, possible, message, [subcommand]) => {
+			if (subcommand === 'list') return undefined;
+			if (!arg && subcommand === 'reset') return undefined;
+			return message.client.arguments.get('string')!.run(arg, possible, message);
+		}
+	],
+	[
+		'rolenames', (arg, possible, message, [subcommand]) => {
+			if (subcommand === 'list' || subcommand === 'reset') return undefined;
+			return message.client.arguments.get('rolenames')!.run(arg, possible, message);
+		}
+	]
+])
```
Is this really the ESLint suggested style? Seems kinda bulky with the [ ] having new lines.

favna replied:
ESLint doesn't nag about it anyway. I like this style because it's readable.

Quantumlyy commented on src/commands/Social/pay.ts:
```diff
@@ -24,24 +26,42 @@ export default class extends SkyraCommand {
-		await user.settings.sync();
+			if (!accepted) return this.denyPayment(message);
 
-		await message.author.decreaseBalance(money);
-		await user.increaseBalance(money);
```
Please implement the logic present in these 2 methods in the new implementation.

Quantumlyy commented on src/commands/Twitch/twitchsubscription.ts:
```diff
@@ -39,51 +38,60 @@ const $KEY = GuildSettings.Notifications.Streams.Twitch.Streamers;
-				// If the subcommand is Show, Reset, or Remove
-				if (
-					type === Type.Show
-					|| type === Type.Reset
-					|| type === Type.Remove
-					// or if the command is Add, the flagArgs include --embed and the status is online then allow no content
-					|| (type === Type.Add && Boolean(message.flagArgs.embed) && status === 0)
-				) return undefined;
-				if (!argument) throw message.language.tget('COMMAND_TWITCHSUBSCRIPTION_REQUIRED_CONTENT');
-				return this.client.arguments.get('...string')!.run(argument, possible, message);
-			});
-	}
+			return message.client.arguments.get('textchannelname')!.run(argument, possible, message);
+		}
+	],
+	[
+		'status', (argument, _, message, [type]) => {
+			if (type === Type.Show || type === Type.Reset) return undefined;
+			if (!argument) throw message.language.tget('COMMAND_TWITCHSUBSCRIPTION_REQUIRED_STATUS');
+
+			const index = message.language.tget('COMMAND_TWITCHSUBSCRIPTION_STATUS_VALUES').indexOf(argument.toLowerCase());
+			if (index === -1) throw message.language.tget('COMMAND_TWITCHSUBSCRIPTION_INVALID_STATUS');
+			return index;
+		}
+	],
+	[
+		'content', (argument, possible, message, [type, , , status]) => {
+			// If the subcommand is Show, Reset, or Remove
+			if (
+				type === Type.Show
+				|| type === Type.Reset
+				|| type === Type.Remove
+				// or if the command is Add, the flagArgs include --embed and the status is online then allow no content
+				|| (type === Type.Add && Boolean(message.flagArgs.embed) && status === 0)
+			) return undefined;
+			if (!argument) throw message.language.tget('COMMAND_TWITCHSUBSCRIPTION_REQUIRED_CONTENT');
+			return message.client.arguments.get('...string')!.run(argument, possible, message);
+		}
+	]
+])
```
[Ditto](https://github.com/skyra-project/skyra/pull/1060#discussion_r453834707)**.**

vladfrangu commented on src/lib/orm/entities/ModerationEntity.ts:
```diff
@@ -0,0 +1,414 @@
+
+	public async invalidate() {
+		if (this.invalidated) return this;
+		const clone = this.clone();
+		try {
+			this.type! |= Moderation.TypeMetadata.Invalidated;
+			await this.save();
+		} catch (error) {
+			this.type = clone.type;
+			throw error;
```
No reason to do an entire clone, just clone the type itself

kyranet replied:
![image](https://user-images.githubusercontent.com/24852502/87346798-6de5fa80-c552-11ea-82a4-1df6a69553dc.png)

vladfrangu commented on src/lib/orm/entities/ModerationEntity.ts:
```diff
@@ -0,0 +1,414 @@
+			.setAuthor(moderator.tag, moderator.displayAvatarURL({ size: 128, format: 'png', dynamic: true }))
+			.setDescription(description)
+			.setFooter(
+				this.#manager.guild.language.tget('MODERATION_LOG_FOOTER', this.caseID),
```
```suggestion
				manager.guild.language.tget('MODERATION_LOG_FOOTER', this.caseID),
```

Quantumlyy commented on src/events/klasaReady.ts:
```diff
@@ -65,4 +69,15 @@ export default class extends Event {
 		}
 	}
 
+	private async initCommandUses() {
+		const { commandCounters } = await DbSet.connect();
+		const entries = await commandCounters.find();
+
+		const missing = this.client.commands
+			.filter(command => !entries.some(entry => entry.id === command.name))
+			.map(command => new CommandCounterEntity().setID(command.name));
+
+		if (missing.length) await commandCounters.insert(missing);
+	}
```
Should this be included seeing as how it's a planned and partially implemented part of #1071.

favna replied:
Two reasons

1. As you say, partially implemented
2. Without we would suddenly not be tracking command uses anymore once this is merged.

favna replied:
Also I don't see any code changes to klasaReady that target command uses in your PR

Quantumlyy commented on src/events/rawChannelDelete.ts:
```diff
@@ -0,0 +1,78 @@
+export default class extends Event {
+
+	public constructor(store: EventStore, file: string[], directory: string) {
+		super(store, file, directory, { name: 'CHANNEL_DELETE', emitter: store.client.ws });
```
Could this possibly be implemented as an Events enum entry?

favna replied:
That's something we can do at a later time and is out-of-scope for this PR tbh. If you want to create a GH issue for it go ahead.

Quantumlyy commented on src/events/rawChannelDelete.ts:
```diff
@@ -0,0 +1,78 @@
+
+export default class extends Event {
+
+	public constructor(store: EventStore, file: string[], directory: string) {
+		super(store, file, directory, { name: 'CHANNEL_DELETE', emitter: store.client.ws });
+	}
```
Is it not the standard t @applyOptions for pieces in this case?

vladfrangu replied:
Can't due to the emitter

vladfrangu commented on src/tasks/delayedGiveawayCreate.ts:
```diff
@@ -0,0 +1,13 @@
+
+	public async run(data: GiveawayCreateData): Promise<PartialResponseValue> {
+		await this.client.giveaways.create(data);
+		return { type: ResponseType.Finished };
```
Finished?

kyranet replied:
This makes Schedule remove the entry from the database.

favna committed:
Update src/lib/orm/entities/ModerationEntity.ts

Co-authored-by: Vlad Frangu <kingdgrizzle@gmail.com>

favna commented on src/tasks/delayedGiveawayCreate.ts:
```diff
@@ -0,0 +1,13 @@
+// This task will be created by the gschedule command when a user schedules a giveaway to be created.
```
`gschedule` command? We don't have such a command

kyranet replied:
Are you saying there's <!-- a multiverse --> no gschedule command?!?

Out of jokes it's `gstart`, which has an alias `giveawayschedule`, but no `gschedule`:
![image](https://user-images.githubusercontent.com/24852502/87347098-f1075080-c552-11ea-8cba-a1da2d9ec844.png)

favna commented on database/Type600.ts:
```diff
@@ -0,0 +1,93 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
```
This needs to be removed before we merge

favna reviewed (changes requested):
<snip>

favna committed:
emit moneytransaction in pay

favna committed:
finish migrations and remove last todo messages for  TORM

favna committed:
assorted changes

favna committed:
really ignore ignored worsd

favna committed:
formatting

favna committed:
fix rpg item entity

favna committed:
fix clear old tables

favna committed:
add user cooldown

Co-authored-by: Antonio Román <kyradiscord@gmail.com>

favna committed:
decorate profile

favna committed:
We should get someone from Purdue to do this. They are the boilerplaters.

favna committed:
throw in queue when not playing and no queue
